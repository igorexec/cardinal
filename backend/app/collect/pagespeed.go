package collect

import (
	"fmt"
	"github.com/go-chi/render"
	"github.com/igorexec/cardinal/app/store"
	"io"
	"log"
	"net/http"
	"time"
)

const googlePageSpeedAPI = "https://www.googleapis.com/pagespeedonline/v5/runPagespeed"

type PageSpeedCollector struct {
	Token string
}

func NewPageSpeedCollector(token string) *PageSpeedCollector {
	return &PageSpeedCollector{Token: token}
}

func (c *PageSpeedCollector) Do(page string) (result store.PageSpeed, err error) {
	log.Printf("[info] pagespeed collector started for %s", page)

	url := fmt.Sprintf("%s?url=%s&key=%s", googlePageSpeedAPI, page, c.Token)
	resp, err := http.Get(url)
	if err != nil {
		log.Printf("[error] failed to collect data for %s: %v", page, err)
		return result, err
	}

	score, err := parsePageSpeed(resp.Body)
	if err != nil {
		log.Fatalf("[error] failed to decode: %v", err)
		return result, err
	}

	log.Printf("[info] pagespeed collector for %s finished", page)

	return store.PageSpeed{
		Score: int(score * 100),
		Page:  page,
		Date:  time.Now().UTC(),
	}, nil
}

func parsePageSpeed(body io.ReadCloser) (float32, error) {
	apiData := struct {
		LighthouseResult struct {
			Categories struct {
				Performance struct {
					Score float32 `json:"score"`
				} `json:"performance"`
			} `json:"categories"`
		} `json:"lighthouseResult"`
	}{}

	if err := render.DecodeJSON(body, &apiData); err != nil {
		return 0, err
	}
	return apiData.LighthouseResult.Categories.Performance.Score, nil
}
