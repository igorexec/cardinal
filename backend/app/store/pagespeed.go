package store

import "time"

type PageSpeed struct {
	ID    string    `json:"id,omitempty" bson:"_id,omitempty"`
	Score int       `json:"score"`
	Page  string    `json:"page"`
	Date  time.Time `json:"time"`
}
