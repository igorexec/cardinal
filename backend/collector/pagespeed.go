package collector

import (
	"fmt"
	"github.com/go-chi/render"
	"github.com/go-pkgz/lgr"
	"github.com/icheliadinski/cardinal/store"
	"io"
	"net/http"
	"time"
)

type PageSpeedAPI struct {
	URL   string
	Token string
}

type Collector struct {
	PageSpeedAPI PageSpeedAPI
}

func (c *Collector) CollectPageSpeed(page string) (result store.PageSpeed, err error) {
	lgr.Printf("[INFO] pagespeed collector started")

	url := fmt.Sprintf("%s?url=%s", c.PageSpeedAPI, page)
	resp, err := http.Get(url)
	if err != nil {
		lgr.Printf("[ERROR] failed to collect data from %s", url)
		return result, err
	}

	score, err := c.mapBodyToScore(resp.Body)
	if err != nil {
		lgr.Printf("[INFO] can't map google API response to score")
		return result, err
	}

	lgr.Printf("[INFO] pagespeed collector finished")

	result = store.PageSpeed{
		Page:  page,
		Score: int(score * 100),
		Date:  time.Now().UTC(),
	}
	return result, nil
}

func (c *Collector) mapBodyToScore(body io.ReadCloser) (float32, error) {
	apiData := struct {
		LighthouseResult struct {
			Categories struct {
				Performance struct {
					Score float32 `json:"score"`
				} `json:"performance"`
			} `json:"categories"`
		} `json:"lighthouseResult"`
	}{}

	if err := render.DecodeJSON(body, &apiData); err != nil {
		return 0, err
	}
	return apiData.LighthouseResult.Categories.Performance.Score, nil
}
