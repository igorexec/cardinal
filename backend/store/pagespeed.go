package store

import "time"

type PageSpeed struct {
	Page  string    `json:"page"`
	Score int       `json:"score"`
	Date  time.Time `json:"date"`
}
