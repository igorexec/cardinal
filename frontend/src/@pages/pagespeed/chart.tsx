import React from 'react';
import {LineChart} from '@ui-kit/charts';

interface Props {
  title: string;
  data?: any;
}

export const PageSpeedChart: React.FC<Props> = ({title}) => {
  const lines = [{
    data: [
      {y: 0, x: new Date('2019-08-11T00:00:00.000')},
      {y: 100, x: new Date('2019-08-12T00:00:00.000')},
      {y: 0, x: new Date('2019-08-13T00:00:00.000')},
      {y: 100, x: new Date('2019-08-14T00:00:00.000')},
      {y: 0, x: new Date('2019-08-15T00:00:00.000')},
      {y: 100, x: new Date('2019-08-16T00:00:00.000')},
    ],
    name: 'www.toryburch.com',
  }, {
    data: [
      {y: 61, x: new Date('2019-08-11T00:00:00.000')},
      {y: 71, x: new Date('2019-08-12T00:00:00.000')},
      {y: 54, x: new Date('2019-08-13T00:00:00.000')},
      {y: 81, x: new Date('2019-08-14T00:00:00.000')},
      {y: 13, x: new Date('2019-08-15T00:00:00.000')},
      {y: 77, x: new Date('2019-08-16T00:00:00.000')},
    ],
    name: 'www.toryburch1.com',
  }];
  return (
    <section>
      <h1>{title}</h1>
      <LineChart data={lines} />
    </section>
  );
};
