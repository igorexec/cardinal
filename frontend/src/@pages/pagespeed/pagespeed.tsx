import React from 'react';
import {PageSpeedChart} from './chart';

export const PageSpeed: React.FC = () => {
  return (
    <div>
      <PageSpeedChart title="Mobile" />
      <PageSpeedChart title="Desktop" />
    </div>
  );
};
