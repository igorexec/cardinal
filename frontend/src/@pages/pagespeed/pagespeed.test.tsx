import React from 'react';
import {shallow, ShallowWrapper} from 'enzyme';
import {PageSpeed} from './pagespeed';

describe('PageSpeed', () => {
  let wrapper: ShallowWrapper;

  beforeAll(() => {
    wrapper = shallow(<PageSpeed />);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
