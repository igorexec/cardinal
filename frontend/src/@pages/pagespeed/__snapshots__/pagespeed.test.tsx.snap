// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSpeed matches snapshot 1`] = `
<div>
  Cardinal [PageSpeed]
</div>
`;
